### PR TITLE
Fixed issue #306: Enforce Meeting.backend_type in Meeting.queue.allowed_backends via API

### PR DESCRIPTION
**Pull request by Frank-duuuu**
Added conditions to `validate()` in `MeetingSerializer`, checking if the `backend_type` of this meeting is in the `allowed_backends` of this queue. Additionally, added unit tests to test the validation logic in `MeetingSerializerTestCase`.

This change ensures that the backend type of a meeting is only allowed if it matches the allowed backends specified for the queue. This prevents incorrect backend types from being associated with meetings and causing issues downstream.

Please review and merge this pull request if it meets the project's standards.

### CODE REVIEW

File: src/officehours_api/serializers.py
Comment (ssciolla) [inline src/officehours_api/serializers.py:271]: Nice work. I'll review this more thoroughly next week. Before that, could you update this validation error message to be more detailed? Something like `f"{backend_type} is not one of the queue's allowed backend types (f{queue.allowed_backends})"`

Comment (Frank-duuuu) [reply src/officehours_api/serializers.py]: Of course! The error message has been updated in the commit ([Changed Error Output](https://github.com/tl-its-umich-edu/remote-office-hours-queue/pull/400/commits/9a04f254686ff5dd59356252b8d904599a31ebb3)).

File: src/officehours_api/serializers.py
Comment (ssciolla) [inline src/officehours_api/serializers.py:272]: So I did a little testing, and I found some issues with this. I think we need to use the same syntax we did for `assignee`, where we check whether it exists using `attrs.get(...)` before we use it. There are a couple ways in the application that a meeting can be `PATCH`ed or updated, and it's not a given that the `backend_type` will be included. Once this is modified, it probably makes sense to add a couple test cases for the patching as well.

Let me know if you need assistance formulating those test cases. There is a PATCH endpoint for meetings in the Swagger UI, and you can get the meeting IDs from the Django admin site or the database.

**Review by ssciolla — CHANGES_REQUESTED** [review]
So I did a little testing, and this needs some modification. Specifically, see my last comment about a bug when the meeting is updated without the `backend_type`. The other comments are mostly about style and code organization -- nice to have but not as important as the bug of course.

File: src/officehours_api/serializers.py
Comment (Frank-duuuu) [reply src/officehours_api/serializers.py]: We have modified the code and added a couple of tests in the test file. We are considering adding a test case (`test_patch_backend_type_invalid()`) in which the meeting is valid first (in person), then we patch it to an invalid type (zoom), and we would get the expected error message. However, since we don't have the credential of zoom type, the error message wouldn't be the one we expected, so the test would not pass until the credential is included. Should we leave the test here or just delete it?

Comment (ssciolla) [reply src/officehours_api/serializers.py]: Looks like you made the right change to the validation. Let me take a look at the tests and I'll let you know if I think they are reasonable. Yes, it's probably a bit weird to test without both backend types enabled.

Comment (Frank-duuuu) [reply src/officehours_api/serializers.py]: I figured out one naive way to solve this is to run the test and set the environment variable at runtime. So it could pass all tests if we just run `docker-compose run -e ZOOM_CLIENT_ID="test" -e ZOOM_CLIENT_SECRET="test" web python manage.py test officehours_api.tests.MeetingSerializerTestCase` by setting the ZOOM environment variables on the go.

**Review by ssciolla — COMMENTED** [review]
Thanks for fixing the bug and adding tests. It looks good. Can you add the `skipIf` decorator to the `zoom` tests and consider moving some of the queue set up to `setUp`? Once that is done, I will approve!

File: src/officehours_api/tests.py
Comment (Frank-duuuu) [reply src/officehours_api/tests.py]: Yeah, using skipIf seems like a valid solution. And thanks for reminding us of the duplicate code issue. We have fixed that in the latest commit (eb34d24).

File: src/officehours_api/tests.py
Comment (ssciolla) [inline src/officehours_api/tests.py:260]: I don't know if this is really testing what happens a meeting is patched. I think we need to something more like what is described here: https://www.django-rest-framework.org/api-guide/serializers/#partial-updates

I think it could be something like this, where you create a model instance and then only pass the updated data.
```
        meeting = Meeting(**{
            'queue': self.inperson_queue.id,
            'attendee_ids': [self.user2.id],
            'agenda': 'test agenda',
            'assignee_id': self.user1.id,
            'backend_type': 'inperson'
        })
        # Patch the meeting to an invalid backend type
        data = {
            'backend_type': 'zoom'
        }
        serializer = MeetingSerializer(meeting, data=data, partial=True)
```
Can we use this approach for the two patch cases?

**Review by ssciolla — COMMENTED** [review]
Thanks for the updates. I think we need to tweak how we're testing partial updates, but otherwise we're good to go.

File: src/officehours_api/tests.py
Comment (Frank-duuuu) [reply src/officehours_api/tests.py]: We tried to achieve something like this but failed to find this partial-updates documents. Thank you so much for the updates! We have added it to our latest commit `d0e34cc`!
For the patch test cases, now we would only assert once after the meeting is patched. So the flow would be to create a new meeting, patch the meeting with new info, and run the validation.

File: src/officehours_api/tests.py
Comment (ssciolla) [inline src/officehours_api/tests.py:220]: Sorry about all the back and forth. Can you specify the attendee in the setup in both these cases? Then maybe just patch the `agenda` in this one, and then the `backend_type` in the other? In the UI, the attendee is basically always set on the initial meeting creation and then never changed. We may as well try to make these realistic or follow real-world usage.

Comment (Frank-duuuu) [reply src/officehours_api/tests.py:220]: Of course! I thought we cannot assign attendees when initialization because when I tried to include `attendee_ids` it would throw an error `Meeting() got an unexpected keyword argument 'attendee_ids'`. But then I realized I forgot to reference the definition of Meeting() in models.py and I just need to include `attendee` from the setup. So now it would work I think!

Comment (ssciolla) [reply src/officehours_api/tests.py]: You're right, I wasn't thinking this through completely. Also apparently I'm a little rusty with the Django ORM. What your doing isn't throwing an error, but it's also not creating an attendee. We have to use special methods to establish this relationship with the other models (`Attendee`, `User`) since it's a [Many-to-Many Relationship](https://docs.djangoproject.com/en/4.2/topics/db/examples/many_to_many/).

I think this is probably the appropriate way to create attendees.

```
        meeting = Meeting(**{
            'queue': self.inperson_queue,
            'agenda': 'test agenda',
            'assignee': self.user1,
            'backend_type': 'inperson'
        })
        meeting.save()
        meeting.attendees.add(self.user2)
        meeting.save()
        logger.info(meeting)
        logger.info(meeting.attendees.all())
```

When debugging this, I recommend adding a logger to the file using `logging.getLogger` and then using `logger.info` to look at the state of the `Meeting` or attendees, like I am above.

This doesn't really affect your tests, but we may as well get the attendee adding right. Could you update these cases one last time please?

Comment (Frank-duuuu) [reply src/officehours_api/tests.py]: Of course, should I define the logger object `logger = logging.getLogger(__name__)` at the beginning of this file or just within those two test case functions?

Comment (ssciolla) [reply src/officehours_api/tests.py]: Beginning of the file. You don't have to commit that though, it's just for your debugging/exploration.

Comment (Frank-duuuu) [reply src/officehours_api/tests.py]: Noted! So I just need to change them into this and it should be fine for commit right?
```
meeting = Meeting(**{
            'queue': self.inperson_queue,
            'agenda': 'test agenda',
            'assignee': self.user1,
            'backend_type': 'inperson'
        })
        meeting.save()
        meeting.attendees.add(self.user2)
        meeting.save()
```

Comment (ssciolla) [reply src/officehours_api/tests.py]: Yep, looks good.

Comment (Frank-duuuu) [reply src/officehours_api/tests.py]: I have pushed the latest commit([52d435a](https://github.com/tl-its-umich-edu/remote-office-hours-queue/pull/400/commits/52d435af99e863f2804a646e6d8627cee024ac23)) to resolve this.

File: src/officehours_api/tests.py
Comment (ssciolla) [inline src/officehours_api/tests.py:173]: I think we should update the language here. The `test_patch_backend_type_none` is not really about setting the backend type to `None`, it's about updating another attribute and making sure everything still works. This was the bug I found before; we were changing the assignee or the agenda on the meeting, but because the backend type was always validated (we didn't have the `if attrs.get("backend_type")` check), we were accessing a key that didn't exist in the partial data we were passing and getting internal server errors.

```suggestion
    is not updated but something else is, and is updated to be invalid.
```

File: src/officehours_api/tests.py
Comment (ssciolla) [inline src/officehours_api/tests.py:210]: In line with the other comment, I think this makes it a bit clearer.
```suggestion
    def test_patch_without_backend_type(self):
```

**Review by ssciolla — APPROVED** [review]
Hi @Frank-duuuu, thanks! I'm approving. Can you review/accept the language changes to the docstring and method name? I think those changes make things a bit clearer. After that, you can click "Rebase and merge" (if it's not available, let me know, I can do it).

File: src/officehours_api/tests.py
Comment (ssciolla) [inline src/officehours_api/tests.py:172]:
```suggestion
    Test the scenarios where the backend type is invalid, is valid,
```

**Comment by ssciolla** [comment]
@Frank-duuuu, done, thanks again for the contribution!

**Pull request merged**
